File: crates/taflow-core/src/stream/tom_de_mark_sequential.rs
//! Stateful Tom DeMark Sequential setup counts.

use std::collections::VecDeque;

/// Buy and sell setup counts for one bar.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct TomDeMarkSequentialValue {
    pub buy: i32,
    pub sell: i32,
}

/// Causal four-bar comparison with capped buy and sell setup counts.
#[derive(Debug, Clone)]
pub struct TomDeMarkSequential {
    closes: VecDeque<f64>,
    buy: i32,
    sell: i32,
    output: TomDeMarkSequentialValue,
    value: Option<TomDeMarkSequentialValue>,
}

impl TomDeMarkSequential {
    /// Creates an empty setup-count state.
    pub fn new() -> Self {
        Self {
            closes: VecDeque::with_capacity(5),
            buy: 0,
            sell: 0,
            output: TomDeMarkSequentialValue { buy: 0, sell: 0 },
            value: None,
        }
    }

    /// Appends one close and returns buy and sell setup counts.
    pub fn append(&mut self, close: f64) -> Option<TomDeMarkSequentialValue> {
        self.closes.push_back(close);
        if self.closes.len() > 5 {
            self.closes.pop_front();
        }
        let result = if self.closes.len() <= 4 {
            None
        } else {
            let comparison = self.closes.front().copied().unwrap();
            if close < comparison {
                self.buy = (self.buy + 1).min(9);
                self.sell = 0;
            } else if close > comparison {
                self.sell = (self.sell + 1).min(9);
                self.buy = 0;
            } else {
                self.buy = 0;
                self.sell = 0;
            }
            Some(TomDeMarkSequentialValue {
                buy: self.buy,
                sell: self.sell,
            })
        };
        self.output = result.unwrap_or(TomDeMarkSequentialValue { buy: 0, sell: 0 });
        self.value = result;
        result
    }

    /// Returns aligned buy/sell outputs for the latest bar.
    pub fn outputs(&self) -> TomDeMarkSequentialValue {
        self.output
    }

    /// Extends close history through the scalar state machine.
    pub fn extend_slice_into(&mut self, close: &[f64], buy: &mut Vec<i32>, sell: &mut Vec<i32>) {
        for &close in close {
            self.append(close);
            let output = self.outputs();
            buy.push(output.buy);
            sell.push(output.sell);
        }
    }

    /// Returns the latest buy/sell setup counts.
    pub fn value(&self) -> Option<TomDeMarkSequentialValue> {
        self.value
    }
    /// Clears setup counts and close history.
    pub fn reset(&mut self) {
        self.closes.clear();
        self.buy = 0;
        self.sell = 0;
        self.output = TomDeMarkSequentialValue { buy: 0, sell: 0 };
        self.value = None;
    }
}

File: python/taflow/indicators/tom_de_mark_sequential.py
"""Canonical native-backed Tom DeMark Sequential adapter."""

from typing import Any

import numpy as np

from .._native import TomDeMarkSequential as _NativeTomDeMarkSequential
from .._series import as_float64_series


class TomDeMarkSequential:
    """Compute causal four-bar buy and sell setup counts.

    ``close`` is required and may be empty for a fresh stream. Rust owns the
    four-bar comparison, capped setup counters, warm-up, and aligned integer
    outputs. ``compute`` returns ``(buy, sell)`` arrays; lifecycle mutators are
    fluent. The independent oracle is ``pandas-ta-classic.td_sequential``.
    """

    def __init__(self, close: Any) -> None:
        self._state = _NativeTomDeMarkSequential()
        self.extend(close)

    def append(self, close: float) -> "TomDeMarkSequential":
        """Append one chronological close and return this adapter."""
        self._state.append(float(close))
        return self

    def extend(self, close: Any) -> "TomDeMarkSequential":
        """Append a converted chronological close history."""
        self._state.extend(as_float64_series(close))
        return self

    def compute(self) -> tuple[np.ndarray, np.ndarray]:
        """Return aligned buy and sell setup arrays."""
        return self._state.compute()

    @property
    def value(self) -> tuple[int, int] | None:
        """Return the latest setup pair, or ``None`` during warm-up."""
        return self._state.value

    def reset(self) -> "TomDeMarkSequential":
        """Reset the state and return this adapter."""
        self._state.reset()
        return self

    def __len__(self) -> int:
        """Return the number of processed close values."""
        return len(self._state)


__all__ = ["TomDeMarkSequential"]
